shai/src/main/scala/at/forsyte/apalache/shai/TransExplorerService.scala

```diff
@@ -95,6 +99,22 @@ class TransExplorerService(connections: Ref[Map[UUID, Conn]], parserSemaphore: S
   // any incoming requests.
   private type RequestConn = Option[Connection]
 
+  // A private object to give us ZIO managed access to an external logger.
+  //
+  // We need to lift the calls to the logger into ZIO effects so they can be
+  // managed by the ZIO schedular for concurrent execution. See the project documentation
+  // in shai/README.md for tips and references on understanding effects in ZIO.
+  private object Log {
+    private def shaiMsg(conn: RequestConn, msg: String): String = {
+      val connId = conn.map(_.id).getOrElse("NO_CONNECTION")
+      s"[Shai:${connId}]: ${msg}"
+    }
+    def debug(conn: RequestConn, msg: String): Result[Unit] = ZIO.effectTotal(logger.debug(shaiMsg(conn, msg)))
+    def info(conn: RequestConn, msg: String): Result[Unit] = ZIO.effectTotal(logger.info(shaiMsg(conn, msg)))
+    def warn(conn: RequestConn, msg: String): Result[Unit] = ZIO.effectTotal(logger.warn(shaiMsg(conn, msg)))
+    def error(conn: RequestConn, msg: String): Result[Unit] = ZIO.effectTotal(logger.error(shaiMsg(conn, msg)))
```

Comment:
Really? I’ve never had a logging framework throw an exception on an attempt to log. What kind of exception do you think these might realistically throw? Do you think we need to defensively handle that?

In any case, ZIO won’t swallow the exceptions, we just get gnarly tracebacks and lose some guarantees if we have an unhandled exception in an effect that should be total.

Reply:
I'm wrong. It will convolute the code to use `effect` here. I don't see any sign in the underlying library or its doc that these are liable to throw errors, so will leave this as `effectTotal` for now! If we end up learning differently, I can ensure they are total.